### DISCUSSION TITLE
Quarkus property quarkus.http.limits.max-header-size

### DISCUSSION THREAD

**Discussion by TilmannBach** [keycloak/keycloak#10236] [Archive]
i need a configuration property for the quarkus properties on defining http limits parameters:

* [quarkus.http.limits.max-header-size](https://quarkus.io/guides/all-config#quarkus-vertx-http_quarkus.http.limits.max-header-size)

**Comment by MaximilianClemens** [comment]
Hello @TilmannBach ,

are you also facing an 431 HTTP Error (Request Header Fields Too Large)?

Regards
Maximilian

**Reply by TilmannBach** [reply]
@MaximilianClemens this is correct, i am facing an 431 HTTP Error response (with NO describing text like `Request Header Fields Too Large`)

**Reply by MaximilianClemens** [reply]
Did you try to put the parameter in `conf/quarkus.properties`? 
I tried it, but it doesn't solve my issue.
([Using unsupported server options](https://www.keycloak.org/server/configuration#_using_unsupported_server_options))

**Reply by TilmannBach** [reply]
yes, we tried this. but it did not helped. you can find my experiments here: https://gitlab.com/clanwars/keycloak-server 
it seems that the large cookies are not the problem because quarkus has a default max size of `20K` and my headers are only `~15K` so maybe there is another component causing the 431 Error!

**Reply by DGuhr** [reply]
Hey. Thanks for reaching out. It would be great if you could provide more information on this case, in terms of setup, how to reproduce, ... -  tbh i never faced the 431 yet, so I am not sure if it's really related to the quarkus distribution or some other component like proxy/lb, as you also state. As you said using the  quarkus options directly seems to have no effect, so also a wrapper on our side would not fix it.

**Comment by Nawarix** [comment]
@TilmannBach did you find a solution to this problem, we are having the same issue when we have multiple clients with one realm

**Comment by edufolly** [comment]
I fixed it adding `QUARKUS_HTTP_LIMITS_MAX_HEADER_SIZE: 40k` environment variable to keycloak container (19.0.3).

I think this happens because the user has too many permissions and the access token is huge.

I'm using keycloak-admin-client and get this error when I try to list all realms.

@MaximilianClemens try to use the environment variable instead of the [quarkus.properties](https://gitlab.com/clanwars/keycloak-server/-/blob/main/conf/quarkus.properties) file.

**Comment by lvian** [comment]
For Kc18, adding `quarkus.http.limits.max-header-size=100K` in the conf/quarkus.properties file solved the issue 👍

**Reply by TCKyle** [reply]
@lvian Hi, I am new to using keycloak. do you mean that just create the new file "conf/quarkus.properties" and add the line "quarkus.http.limits.max-header-size=100K", it can be fixed?

**Reply by lvian** [reply]
Yes, if you are using the Quarkus distribution, you will have the `conf/` folder, which initially only has the keycloak.conf file. You can then create the `quarkus.properties` file in there an add quarkus specific configuration.

After that a `./bin/kc.sh` build and `./bin/kc.sh` start-dev should do the trick.

**Reply by mihkelpoldoja** [reply]
Should this quarkus.properties file contain all other values (like db username, password etc.) also, needed for running Keycloak?
Because then I create this file, then starting Keycloak server fails with complaining about wrong password. Although I have specified db password and username from command line parameters: --db-username=postgres --db-password=asdf --db=postgres 
And I specify username and password in quarkus.properties file, it still doesn't work. 
My quarkus.properties file:
quarkus.http.limits.max-header-size=100K
quarkus.datasource.password=asdf
quarkus.datasource.username=postgres

**Reply by sschu** [reply]
@mihkelpoldoja No, you should only specify the additional quarkus properties in the quarkus.properties file i.e. the quarkus.http.limits.max-header-size. Database configuration is via Keycloak properties. Works fine for us.

**Reply by aazaroual** [reply]
I had the same 431 error in Keycloak 24.0.5 and I solved it by creating /opt/keycloak/24.0.5/conf/quarkus.properties and putting the following lines in it:
```
quarkus.http.limits.max-header-list-size=16384
quarkus.http.limits.max-header-size=100K
```
Restart Keycloak and now it works! **BUT BE CAREFUL: BOTH LINES MUST BE THERE, IF YOU PUT ONLY ONE OF THOSE IT WON'T WORK!**

**Comment by gypark** [comment]
Hello, Is there any update about this issue?

I am trying to run Keycloak 24.0.1 and now I am facing the same issue.
It's not the problem of keycloak, but of my company which use so many (and big) cookies. But I can not handle them so I have to make keyclock accept larger headers.

I tried both conf/quarkus.properties and an environment variable above. But they didn't work.

---

Edit:

After several tries, I found that `quarkus.http.http2=false` works in my case.

FYI:
* Ubuntu 22.04
* docker-ce 24.0.5
* keycloak image based on `quay.io/keycloak/keycloak:24.0.1`
* `conf/keycloak.conf`
```
db=mssql
db-username=***
db-password=***
# db-url is provided via command line arguments in a docker compose file


transaction-xa-enabled=false
health-enabled=true
metrics-enabled=true

https-certificate-file=/opt/ssl/***.pem
https-certificate-key-file=/opt/ssl/***.pem
https-port=443

hostname=***
```
* `conf/quarkus.properties`
```
quarkus.datasource.user-store.db-kind=mssql
quarkus.datasource.user-store.username=***
quarkus.datasource.user-store.password=***
quarkus.datasource.user-store.jdbc.url=***
quarkus.log.level=TRACE
quarkus.log.handler=console
quarkus.log.min-level=TRACE
# These two lines were not effective
# quarkus.http.limits.max-header-size=400K
# quarkus.http.limits.max-chunk-size=32000

# effective, even without above options about max-header-size
quarkus.http.http2=false
```

**Reply by shawkins** [reply]
@gypark please open an issue or start a new thread if you believe you have a situation where the quarkus setting either via the quarkus.properties or env is not effective. There were confirmations above that this did work for users, so there could be something else going on in your situation.

**Reply by gypark** [reply]
@shawkins I found another property that was effective and I updated my comment. Thanks.

**Comment by veerakishorekallam** [comment]
Hi @gypark 

We are using Keycloak 24.0.3 (FROM quay.io/keycloak/keycloak:24.0.3) and added the quarkus.properties under /opt/keycloak/conf/  but we are getting token too large error and only it allows 8KB header size.

Please suggest.

Thanks
Veera

**Reply by shawkins** [reply]
#29770 experienced a similar issue - the problem was with their proxy, not keycloak.

**Comment by Tirupathi-LLC** [comment]
curl -X 'GET' \
  'http://0.0.0.0:9010/pharmacyData/getPharmacyList?IsFaxPharmacyNeeded=yes&PharmacyId=here i am sending 800 or 1000 pharamcy Id's&limit=200&loadInactivePharmacies=false&offset=0&orderById=false' \
  -H 'accept: application/json'
  


414 Undocumented | Error: Request-URI Too Long

can any one give the solution i am doing oin quarkus

**Reply by shawkins** [reply]
@Tirupathi-LLC this is a Keycloak discussion, please move your question to https://github.com/quarkusio/quarkus/discussions
